src/HashBase.ts

```diff
@@ -1,4 +1,142 @@
+import { Disklet } from 'disklet'
+
+import { BaseletConfig, BaseType } from './types'
+
 export interface HashBase {
   insert(partition: string, hash: string, data: any): Promise<unknown>
   query(partition: string, hashes: string[]): Promise<any[]>
 }
+
+interface HashBaseConfig extends BaseletConfig {
+  prefixSize: number
+}
+
+interface BucketDictionary {
+  [bucketName: string]: {
+    bucketFetcher: Promise<void>
+    bucketData: {
+      [hash: string]: any
+    }
+  }
+}
+
+export function openHashBase(
+  disklet: Disklet,
+  databaseName: string
+): Promise<HashBase> {
+  // check that the db exists and is of type HashBase
+
+  function getConfig(): Promise<HashBaseConfig> {
+    return disklet
+      .getText(`${databaseName}/config.json`)
+      .then(serializedConfig => JSON.parse(serializedConfig))
+  }
+
+  return getConfig().then(configData => {
+    return {
+      insert(
+        partition: string = '/',
+        hash: string,
+        data: any
+      ): Promise<unknown> {
+        // check that partition only contains letters, numbers, and underscores
+        // if no partition, then root
+        const { prefixSize } = configData
+
+        if (typeof hash !== 'string' || hash.length < prefixSize) {
+          return Promise.reject(
+            new Error(`hash must be a string of length at least ${prefixSize}`)
+          )
+        }
+
+        const prefix = hash.substring(0, prefixSize)
+        const bucketFilename = `${prefix}.json`
+        const bucketPath = `${databaseName}/${partition}/${bucketFilename}`
+        return disklet
+          .list(`${bucketPath}`)
```

Comment:
No need to check if the file exists. Just try to read the file and if it fails, catch the error and just write to the file